File: packages/shared/src/tournament.ts
export type TournamentType = "round_robin" | "knockout" | "groups" | "repechage";

export interface Tournament {
  id: string;
  title: string;
  subtitle: string | null;
  image_url: string | null;
  type: TournamentType;
  kind: "beach" | "indoor";
  sex: "male" | "female";
  slots: number;
  num_groups: number;
  teams_per_group: number;
  advance_per_group: number;
  prize_silver_1: number;
  prize_silver_2: number;
  prize_silver_3: number;
  prize_gold_1: number;
  prize_gold_2: number;
  prize_gold_3: number;
  status: "open" | "running" | "finished";
  entry_count: number;
  team_size: number;
}

export interface TournamentEntry {
  id: string;
  user_id: string;
  team_name: string;
  athlete_ids: string[];
  group_no: number | null;
  points: number;
  wins: number;
  losses: number;
  sets_won: number;
  sets_lost: number;
  placement: number | null;
}

export interface TournamentMatch {
  id: string;
  stage: string;
  group_no: number | null;
  round_no: number;
  order: number;
  entry_a_id: string | null;
  entry_b_id: string | null;
  a_name: string;
  b_name: string;
  score_a: number | null;
  score_b: number | null;
  winner_entry_id: string | null;
  status: string;
}

export interface TournamentDetail {
  tournament: Tournament;
  entries: TournamentEntry[];
  matches: TournamentMatch[];
  my_entry_id: string | null;
  athletes: Record<string, import("./athlete").Athlete>;
}

export const TOURNAMENT_TYPE_LABEL: Record<TournamentType, string> = {
  round_robin: "Pontos Corridos",
  knockout: "Mata-mata",
  groups: "Chaves de Grupo",
  repechage: "Repescagem",
};

export const TOURNAMENT_STATUS_LABEL: Record<string, string> = {
  open: "Inscrições abertas",
  running: "Em andamento",
  finished: "Encerrado",
};

/** Nome legível da fase de uma partida (Oitavas/Quartas/Semis/Final/3º…). */
export function matchRoundLabel(m: TournamentMatch, matches: TournamentMatch[]): string {
  if (m.stage === "rr") return "Pontos corridos";
  if (m.stage === "group") return `Grupo ${m.group_no ?? ""}`.trim();
  if (m.stage === "bronze") return "Disputa de 3º lugar";
  const rep = m.stage === "rep" || m.stage === "rep_final";
  const koStages = rep ? ["rep", "rep_final"] : ["ko", "final"];
  const totalRounds = matches
    .filter((x) => koStages.includes(x.stage))
    .reduce((mx, x) => Math.max(mx, x.round_no), 0);
  const fromEnd = totalRounds - m.round_no; // 0 = final, 1 = semi, 2 = quartas…
  const prefix = rep ? "Repescagem — " : "";
  const names: Record<number, string> = {
    0: "Final",
    1: "Semifinais",
    2: "Quartas de final",
    3: "Oitavas de final",
    4: "16-avos de final",
  };
  return prefix + (names[fromEnd] ?? `Rodada ${m.round_no}`);
}

File: apps/api/app/schemas/tournament.py
"""Schemas (DTOs) do módulo de Competições."""

import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.athlete import AthleteOut


class TournamentOut(BaseModel):
    id: uuid.UUID
    title: str
    subtitle: str | None = None
    image_url: str | None = None
    type: str
    kind: str
    sex: str
    slots: int
    num_groups: int
    teams_per_group: int
    advance_per_group: int
    prize_silver_1: int
    prize_silver_2: int
    prize_silver_3: int
    prize_gold_1: int
    prize_gold_2: int
    prize_gold_3: int
    status: str
    entry_count: int
    team_size: int


class TournamentCreate(BaseModel):
    title: str = Field(min_length=1, max_length=80)
    subtitle: str | None = Field(default=None, max_length=120)
    image_url: str | None = None
    type: Literal["round_robin", "knockout", "groups", "repechage"] = "round_robin"
    kind: Literal["beach", "indoor"] = "beach"
    sex: Literal["male", "female"] = "male"
    slots: int = Field(default=8, ge=2, le=64)
    num_groups: int = Field(default=2, ge=1, le=16)
    teams_per_group: int = Field(default=4, ge=2, le=16)
    advance_per_group: int = Field(default=2, ge=1, le=8)
    prize_silver_1: int = Field(default=0, ge=0)
    prize_silver_2: int = Field(default=0, ge=0)
    prize_silver_3: int = Field(default=0, ge=0)
    prize_gold_1: int = Field(default=0, ge=0)
    prize_gold_2: int = Field(default=0, ge=0)
    prize_gold_3: int = Field(default=0, ge=0)


class TournamentEntryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    team_name: str
    athlete_ids: list[str] = []
    group_no: int | None = None
    points: int
    wins: int
    losses: int
    sets_won: int
    sets_lost: int
    placement: int | None = None


class TournamentMatchOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    stage: str
    group_no: int | None = None
    round_no: int
    order: int
    entry_a_id: uuid.UUID | None = None
    entry_b_id: uuid.UUID | None = None
    a_name: str
    b_name: str
    score_a: int | None = None
    score_b: int | None = None
    winner_entry_id: uuid.UUID | None = None
    status: str


class TournamentDetailOut(BaseModel):
    tournament: TournamentOut
    entries: list[TournamentEntryOut]
    matches: list[TournamentMatchOut]
    my_entry_id: uuid.UUID | None = None
    # Atletas inscritos (por id) para ver o card de cada um na gestão do torneio.
    athletes: dict[str, AthleteOut] = {}


class RegisterRequest(BaseModel):
    athlete_ids: list[uuid.UUID] = Field(min_length=1, max_length=6)


class MatchResultRequest(BaseModel):
    score_a: int = Field(ge=0, le=5)
    score_b: int = Field(ge=0, le=5)
